### PR TITLE
Generate issues for out of date doc entries 

### PR DESCRIPTION
**Pull request by pipo02mix**
### What does this PR do?

We want to create issues in `giantswarm/giantswarm` repository once the doc page is out of date (passed a year without review). 

Towards https://github.com/giantswarm/giantswarm/issues/18329

(Please set a descriptive PR title. Use this space for additional explanations.)

### What does it look like?

- It modifies the python validation script to add the option of parsing results as json. 
- It creates a Github Action that runs that script and creates issues, when they are not existing yet, to review the outdated pages

(If this is more than a textual change, a screenshot can help in some cases to speed up the review process.)

### Any background context you can provide?

(Please link public issues or summarize if not public.)

### Have you maintained the front matter?

(Please bump the last_review_date in case this qualifies as a review for an entire page. Provide user_questions which should be answered in the page. Provide a meaningful description.)

### CODE REVIEW

File: scripts/validate-front-matter/script.py [hunk 1]
```diff
@@ -340,6 +367,8 @@ def format_message_part(check_id, severity_str, description, value):
             'line': 1,
             'end_line': end_line,
             'title': headline,
+            'owner': check.get('owner'),
+            'page_title': check.get('page_title'),
```
Comment (pipo02mix) [inline scripts/validate-front-matter/script.py:371]:
```suggestion
```

File: scripts/validate-front-matter/script.py [hunk 2]
```diff
@@ -577,23 +606,31 @@ def validate(content, fpath, validation):
...
+                    'page_title': fm['title'],
+                    'owner': fm['owner'],
                     'value': fm['last_review_date'],
                 })
         else:
             if not ignored_path(fpath, NO_LAST_REVIEW_DATE):
                 result['checks'].append({
                     'check': NO_LAST_REVIEW_DATE,
+                    'page_title': fm['title'],
+                    'owner': fm['owner'],
```
Comment (pipo02mix) [inline scripts/validate-front-matter/script.py:633]:
```suggestion
                        'value': fm['last_review_date'],
                    })
                elif diff < datetime.timedelta(seconds=0):
                    # in the future
                    result['checks'].append({
                        'check': INVALID_LAST_REVIEW_DATE,
                        'value': fm['last_review_date'],
                    })
            else:
                result['checks'].append({
                    'check': INVALID_LAST_REVIEW_DATE,
                    'value': fm['last_review_date'],
                })
        else:
            if not ignored_path(fpath, NO_LAST_REVIEW_DATE):
                result['checks'].append({
                    'check': NO_LAST_REVIEW_DATE,
```

File: scripts/validate-front-matter/script.py [hunk 3]
```diff
@@ -574,7 +601,11 @@ def validate(content, fpath, validation):
         if 'last_review_date' in fm:
             if type(fm['last_review_date']) is datetime.date:
                 diff = todays_date - fm['last_review_date']
-                if diff > datetime.timedelta(days=365):
+                expiration = 365
```
Comment (pipo02mix) [inline scripts/validate-front-matter/script.py:604]: Now we support a new metadata field to specify the expiration in days, requested by @lyind

File: .github/workflows/generate_review_issues.yaml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        uses: actions/checkout@v3
+
+      - name: Install dependencies
+        run: pip3 install click colored
+
+      - name: Validate front matter for changed files
+        run: |
+          git fetch --no-tags origin "${GITHUB_BASE_REF}":"${GITHUB_BASE_REF}"
+          git diff --name-only "refs/heads/${GITHUB_BASE_REF}" -- . | tee files.txt
+          echo "ISSUES=$(cat files.txt | python3 scripts/validate-front-matter/script.py --validation last-reviewed --json)" >> $GITHUB_ENV
```
Comment (carillan81) [inline .github/workflows/generate_review_issues.yaml:23]: shouldn't this be "--output json"?

Comment (pipo02mix) [reply .github/workflows/generate_review_issues.yaml:23]: good catch

File: .github/workflows/generate-review-issues.yaml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+          git diff --name-only "refs/heads/${GITHUB_BASE_REF}" -- . | tee files.txt
+          echo "ISSUES=$(cat files.txt | python3 scripts/validate-front-matter/script.py --validation last-reviewed --output json)" >> $GITHUB_ENV
+      
+      - name: Generate issues of out of date docs
+        uses: giantswarm/open-issue@0.1.0
+        with: 
+          token: ${{ secrets.ISSUE_AUTOMATION }}
+          org: giantswarm
+          repo: giantswarm
+          issues: ${{ env.ISSUES }}
```
Comment (marians) [inline .github/workflows/generate-review-issues.yaml:30]: Trailing newline

```suggestion
          issues: ${{ env.ISSUES }}

```

**Comment by marians** [comment]
Looks sane.

> It creates a Github Action that runs that script and creates issues, when they are not existing yet.

How is the existence checked, to avoid duplicates?

**Comment by pipo02mix** [comment]
> How is the existence checked, to avoid duplicates?

It uses a convention for the title and before creating new one, it searches if there is already one with that name

File: scripts/validate-front-matter/script.py (hunk 3 above)
Comment (lyind) [reply scripts/validate-front-matter/script.py:608]: Thank you, LGTM!

We need to test this.

**Review by lyind — APPROVED** [review]
LGTM

Would suggest testing this well, as it's somewhat complex.

Thanks to @marians for asking about the prevention of duplicate issue creations.

**Comment by pipo02mix** [comment]
@marians any concern? can I get your approval?

**Comment by lyind** [comment]
Missed it during review, but: https://github.com/giantswarm/giantswarm/issues/18329#issuecomment-1448215289

**Pull request merged**